blog post about QuestDB and crypto (#801)
questdb/questdb.io: blog/2022-04-04-query-a-live-crypto-data-stream-in-millisecs.md

rows dataset of historical taxi rides in NYC <sup>[1]</sup> and a geospatial
dataset that contains the locations of 250k unique ships <sup>[2]</sup>
overtime. Today, we introduce a new dataset on the same demo instance: crypto
market data ingested in real-time from the Coinbase Exchange. For ingestion, we
used a convenient Python library
[Cryptofeed](https://github.com/bmoscon/cryptofeed), a cryptocurrency exchange
feed handler that supports QuestDB. And for visualization, we used Grafana to
create the respective charts.

<!--truncate-->

We ingest the following into QuestDB in real-time for each BTC-USD and ETH-USD
trades coming through the exchange Coinbase:

- price
- side (buy/sell)
- amount

To get you started, we added a set of example queries in the live demo of
QuestDB web console. These pre-written queries leverage the standard SQL syntax
and time-series SQL extensions in QuestDB. When clicking on a query, it's
automatically added to the SQL editor. Then, click the &quot;Run&quot; button or
press <kbd>F9</kbd> to execute the query.

Let's go through these sample queries one by one.  

## Last prices of BTC and ETH

> To find out the latest prices of BTC and ETH in USD. We use the syntax

Reviewer: ```suggestion
To find out the latest prices of BTC and ETH in USD. We use the
```

> `LATEST BY`, which is native to QuestDB's SQL Engine.

Reviewer: ```suggestion
[`LATEST ON`](https://questdb.io/docs/reference/sql/latest-on) syntax, which is native to QuestDB's SQL Engine.
```

Click

> [here](https://demo.questdb.io/?query=select%20%2a%20from%20trades%0Alatest%20by%20symbol%20%0AWHERE%20symbol%20in%20%28%27BTC-USD%27%2C%20%27ETH-USD%27%29%3B)

Reviewer: This query should be:
```sql
SELECT *
FROM trades
WHERE symbol in ('BTC-USD', 'ETH-USD')
LATEST ON timestamp PARTITION BY symbol;
```

Reviewer 2: @puzpuzpuz Thanks! @argshook guess we should also update the example query in the demo box?

Author: Thanks! Updated the URL with suggested query.
The `Example Queries` popup should also be updated, that effort is tracked separately.

Marking as resolved, please reopen if there's anything left

Reviewer: Thanks!

QuestDB web console. These pre-written queries leverage the standard SQL syntax
and time-series SQL extensions in QuestDB. When clicking on a query, it's
automatically added to the SQL editor. Then, click the <kbd>Run</kbd> button or
press <kbd>F9</kbd> to execute the query. Despite the large amount of data
stored on the demo instance, the queries should come back in milliseconds!

Let's go through these sample queries one by one.

## Last prices of BTC and ETH

To find out the latest prices of BTC and ETH in USD. We use the
[`LATEST ON`](https://questdb.io/docs/reference/sql/latest-on) syntax, which is
native to QuestDB's SQL Engine:

```questdb-sql title="Latest BTC and ETH prices"
SELECT * FROM trades
WHERE symbol in ('BTC-USD', 'ETH-USD')
LATEST ON timestamp PARTITION BY symbol;
```

Click [here](https://demo.questdb.io/?query=SELECT%20%2a%0AFROM%20trades%0AWHERE%20symbol%20in%20%28%27BTC-USD%27%2C%20%27ETH-USD%27%29%0ALATEST%20ON%20timestamp%20PARTITION%20BY%20symbol%3B)
to fill the pre-written query directly in the SQL editor, then run the query.

Below is the chart for Bitcoin and Ethereum prices with a time sample of 10
seconds:

<iframe

>   src="https://dashboard.questdb.io/d-solo/624FG0snk/public-dashboard-1?orgId=1&panelId=10"

Author: ```suggestion
  src="https://dashboard.questdb.io/d-solo/624FG0snk/public-dashboard-1?orgId=1&panelId=10&refresh=5s"
```

volumes with 15-minute intervals. We use the
[`SAMPLE BY`](https://questdb.io/docs/reference/sql/sample-by/) syntax, which
aggregates time series data into homogeneous time chunks:

```questdb-sql title="Candle chart with 15-minute intervals"
SELECT timestamp,
       first(price) AS open,
       last(price) AS close,
       min(price),
       max(price),
       sum(amount) AS volume
FROM trades
WHERE symbol = 'BTC-USD' AND timestamp > dateadd('d', -1, now())
SAMPLE BY 15m ALIGN TO CALENDAR;
```

Click
[here](https://demo.questdb.io/?query=SELECT%20timestamp%2C%20first%28price%29%20as%20open%2C%20last%28price%29%20as%20close%2C%20min%28price%29%2C%20max%28price%29%2C%20sum%28amount%29%20as%20volume%0AFROM%20trades%0AWHERE%20symbol%20%3D%20%27BTC-USD%27%20and%20timestamp%20%3E%20dateadd%28%27d%27%2C%20-1%2C%20now%28%29%29%0ASAMPLE%20BY%2015m%20ALIGN%20TO%20CALENDAR%20WITH%20OFFSET%20%2700%3A00%27%3B)
to fill the pre-written query directly in the SQL editor, then run the query.

This real-time chart on Grafana plots the candle chart with a time sample of 10
seconds. We also show the volume traded on a secondary axis.

<iframe

>   src="https://dashboard.questdb.io/d-solo/624FG0snk/public-dashboard-1?orgId=1&panelId=8"

Author: ```suggestion
  src="https://dashboard.questdb.io/d-solo/624FG0snk/public-dashboard-1?orgId=1&panelId=8"
```

for the volume of trades during that period. This query includes the
[`WHERE`](https://questdb.io/docs/reference/sql/where/) clause that is
accelerated by our new JIT Compiler (see the _lightning_ in the logs). And once
again, we downsample the dataset using `SAMPLE BY`.

```questdb-sql title="Volume-weighted average price"
SELECT timestamp,
       sum(price * amount) / sum(amount) AS vwap_price,
       sum(amount) AS volume
FROM trades
WHERE symbol = 'BTC-USD' AND timestamp > dateadd('d', -1, now())
SAMPLE BY 15m ALIGN TO CALENDAR;
```

Click
[here](https://demo.questdb.io/?query=SELECT%20timestamp%2C%20sum%28price%20%2a%20amount%29%20%2F%20sum%28amount%29as%20vwap_price%2C%20sum%28amount%29%20as%20volume%20%0AFROM%20trades%0AWHERE%20symbol%20%3D%20%27BTC-USD%27%20and%20timestamp%20%3E%20dateadd%28%27d%27%2C%20-1%2C%20now%28%29%29%0ASAMPLE%20BY%2015m%20ALIGN%20TO%20CALENDAR%20WITH%20OFFSET%20%2700%3A00%27%3B)
to fill the pre-written query directly in the SQL editor, then run the query.

The following chart displays the distribution of trades based on their size and
paints a more granular picture of volume traded.

<iframe

>   src="https://dashboard.questdb.io/d-solo/624FG0snk/public-dashboard-1?orgId=1&panelId=12"

Author: ```suggestion
  src="https://dashboard.questdb.io/d-solo/624FG0snk/public-dashboard-1?orgId=1&panelId=12&refresh=5s"
```

to fill the pre-written query directly in the SQL editor, then run the query.

```questdb-sql title="Implied BTC-ETH exchange rate"
WITH btc AS (
    SELECT timestamp, price
    FROM trades
    WHERE symbol = 'BTC-USD' AND timestamp > dateadd('d', -30, now())
), eth AS (
    SELECT timestamp, price
    FROM trades
    WHERE symbol = 'ETH-USD' and timestamp > dateadd('d', -30, now())
)
SELECT btc.timestamp btc_time, btc.price btc_price, eth.price eth_price, round(btc.price/eth.price, 3) btc_to_eth_ratio
FROM btc
ASOF JOIN eth;
```

The following chart plots three series: BTC-USD, ETH-USD and the implied BTC-ETH cross price.

<iframe

>   src="https://dashboard.questdb.io/d-solo/624FG0snk/public-dashboard-1?orgId=1&panelId=14"

Author: ```suggestion
  src="https://dashboard.questdb.io/d-solo/624FG0snk/public-dashboard-1?orgId=1&panelId=14&refresh=5s"
```

Despite the large amount of data stored on the AWS instance, the queries should come back in milliseconds. QuestDB is fast for time-based queries and has been optimized around that from its inception. QuestDB stores data in chronological order, and deals with un-ordered data with on the fly live re-ordering of data before landing on disk. Data is also column based and partitioned by time; only the relevant time partitions and columns will be lifted for each query.

This time, we also introduce live Grafana dashboards on top of the data being streamed into QuestDB, and display them in this post below.

## Last prices of BTC and ETH

This query returns the latest price of both BTC-USD and ETH-USD. We use the syntax `LATEST BY` native to QuestDB&#39;s SQL.

Click [here](https://demo.questdb.io/?query=select%20%2a%20from%20trades%0Alatest%20by%20symbol%20%0AWHERE%20symbol%20in%20%28%27BTC-USD%27%2C%20%27ETH-USD%27%29%3B) to fill the query directly on our SQL console with live data.

![](/img/blog/2022-04-07/last-prices-of-btc-and-eth-screenshot.png)

> Below is the chart for Bitcoin and Ethereum prices with a time sample of 10 seconds

Author: @tris0laris wdyt?

```suggestion
Below is the chart for Bitcoin and Ethereum prices with a time sample of 10 seconds. It is live Grafana chart you can interact with.
```

Reviewer 2: @argshook I included info about Grafana in the intro paragraphs. wdyt?

> <iframe src="https://dashboard.questdb.io/d-solo/624FG0snk/public-dashboard-1?orgId=1&panelId=10" width="100%" height="300" frameborder="0"></iframe>

Author: @tris0laris wdyt, should grafana charts have auto refresh?

```suggestion
<iframe src="https://dashboard.questdb.io/d-solo/624FG0snk/public-dashboard-1?orgId=1&panelId=10&refresh=10s" width="100%" height="300" frameborder="0"></iframe>
```
